File: src/linked_lists/java/Partition.java
package linked_lists.java;

import static linked_lists.java.Node.printLinkedList;
import static linked_lists.java.RemoveDups.initLinkedList;

import java.util.Scanner;

/**
 *
 */
public class Partition {

  public static void main(String[] args) {
    Node list = initLinkedList(10);
    System.out.println("Linked list:");
    printLinkedList(list);

    System.out.println("Input an integer:");
    Scanner in = new Scanner(System.in);
    int     n  = in.nextInt();
    System.out.println("List after partition:");
    printLinkedList(partition(list, n));
  }

  private static Node partition(Node head, int x) {
    if (head == null) {
      return null;
    }

    Node prev = head;
    Node ptr  = head.next;
    while (ptr != null) {
      if (ptr.data < x) {
        prev.next = ptr.next;

        ptr.next = head;
        head = ptr;

        ptr = prev.next;
      } else {
        prev = ptr;
        ptr = ptr.next;
      }
    }

    return head;
  }
}

File: src/linked_lists/kotlin/Partition.kt
package linked_lists.kotlin

/**
 *
 */
fun main(args: Array<String>) {
    println("Linked list:")
    val list = initLinkedList(10)
    printLinkedList(list)

    println("Input the number:")
    val n = readLine().toString().toInt()
    println("List after partition:")
    printLinkedList(partition(list, n))
}

fun partition(head: Node, x: Int): Node {
    var h = head
    var prev = head
    var ptr = head.next
    while (ptr != null) {
        if (ptr.data < x) {
            prev.next = ptr.next
            ptr.next = h
            h = ptr
            ptr = prev.next
        } else {
            prev = ptr
            ptr = ptr.next
        }
    }

    return h
}